File: plan-wise-vue-python/backend/services/income_service.py
from sqlalchemy import and_
from datetime import datetime
from app.database import SessionLocal, models
from util.date_data_generator import group_by_date

class IncomeService:
    def __init__(self):
        self.session = SessionLocal()

    def create_income(self, user_id, income_request):
        user = self.session.query(models.PwUser).filter_by(id=user_id).first()
        if not user:
            return {"success": False,"message": "User not found"}

        income = models.PwIncome(
            user_id=user_id,
            source=income_request.source,
            amount=income_request.amount,
            start_date=income_request.start_date,
            frequency=income_request.frequency
        )
        self.session.add(income)
        self.session.commit()
        income_id = income.id

        budget_data = group_by_date([{
            "source": income_request.source,
            "income_id": income_id,
            "amount": income_request.amount,
            "start_date": income_request.start_date,
            "frequency": income_request.frequency
        }])
        budget_incomes = []
        for date, incomes in budget_data.items():
            for inc in incomes:
                budget_income = models.PwBudgetTableIncome(
                    date=date,
                    user_id=user_id,
                    income_id=income_id,
                    source=inc["source"],
                    amount=inc["amount"],
                    start_date=inc["start_date"],
                    frequency=inc["frequency"]
                )
                budget_incomes.append(budget_income)

        self.session.add_all(budget_incomes)
        self.session.commit()

        return {"success": True, "message": "Income and budget incomes created successfully", "income": income.as_dict()}

    def get_income(self, user_id, income_id):

        user = self.session.query(models.PwUser).filter_by(id=user_id).first()
        if not user:
            return { "success": False, "message": "User not found"}

        income = self.session.query(models.PwIncome).filter_by(id=income_id, user_id=user_id).first()
        if not income:
            return { "success": True, "message": "Income not found or not owned by the user"}

        return {"success": True, "message": "Income retrieved successfully", "income": income}

    def get_income_by_user(self, user_id):
        user = self.session.query(models.PwUser).filter_by(id=user_id).first()
        if not user:
            return { "success": False, "message": "User not found"}
        income = self.session.query(models.PwIncome).filter_by(user_id=user_id).order_by(models.PwIncome.start_date).all()
        if not income:
            return { "success": False, "message": "Income not found or not owned by the user"}

        return {"success": True, "message": "Income retrieved successfully", "income": income}

    def update_income(self, user_id, income_id, income_request):
        user = self.session.query(models.PwUser).filter_by(id=user_id).first()
        if not user:
            return { "success": False, "message": "User not found"}

        income = self.session.query(models.PwIncome).filter_by(id=income_id, user_id=user_id).first()
        if not income:
            return { "success": False, "message": "Income not found or not owned by the user"}

        income.source = income_request.source if income_request.source is not None else income.source
        income.amount = income_request.amount if income_request.amount is not None else income.amount
        income.start_date = income_request.start_date if income_request.start_date is not None else income.start_date
        income.frequency = income_request.frequency if income_request.frequency is not None else income.frequency

        self.session.commit()

        self.session.query(models.PwBudgetTableIncome).filter(
            models.PwBudgetTableIncome.date > str(datetime.now().date()),
            models.PwBudgetTableIncome.income_id == income_id, 
            models.PwBudgetTableIncome.user_id == user_id
        ).delete(synchronize_session=False)
        self.session.commit()

        budget_data = group_by_date([{
            "source": income_request.source,
            "income_id": income_id,
            "amount": income_request.amount,
            "start_date": income_request.start_date,
            "frequency": income_request.frequency
        }])

        budget_incomes = []
        for date, incomes in budget_data.items():
            if datetime.now().date() > datetime.strptime(date,'%Y-%m-%d').date():
                continue
            for inc in incomes:
                budget_income = models.PwBudgetTableIncome(
                    date=date,
                    user_id=user_id,
                    income_id=income_id,
                    source=inc["source"],
                    amount=inc["amount"],
                    start_date=inc["start_date"],
                    frequency=inc["frequency"]
                )
                budget_incomes.append(budget_income)

        self.session.add_all(budget_incomes)
        self.session.commit()

        return { "success": True, "message": "Income updated successfully", "income": income}

    def delete_income(self, user_id, income_id):
        user = self.session.query(models.PwUser).filter_by(id=user_id).first()
        if not user:
            return { "success": False, "message": "User not found"}

        income = self.session.query(models.PwIncome).filter_by(id=income_id, user_id=user_id).first()
        if not income:
            return { "success": False, "message": "Income not found or not owned by the user"}

        self.session.delete(income)
        self.session.commit()
        self.session.query(models.PwBudgetTableIncome).filter(
            models.PwBudgetTableIncome.date > str(datetime.now().date()),
            models.PwBudgetTableIncome.income_id == income_id, 
            models.PwBudgetTableIncome.user_id == user_id
        ).delete(synchronize_session=False)
        self.session.commit()

        return { "success": True, "message": "Income deleted successfully"}

File: plan-wise-react-node/backend/services/income.service.js
const db = require('./db.service');
const { group_by_date } = require('../util/date_data_generator');
const Sequelize = require("sequelize");

class IncomeService {
    constructor() {
        // This is intentional
    }

    async create_income(user_id, income_request) {
        const user = await db.pw_users.findOne({ where: { id: user_id } });
        if (!user) {
            return { success: false, message: "User not found" };
        }

        let income = await db.pw_income.create({
            user_id: user_id,
            source: income_request.source,
            amount: income_request.amount,
            start_date: income_request.start_date,
            frequency: income_request.frequency
        });

        let income_id = income.id;

        let budget_data = group_by_date([{
            source: income_request.source,
            income_id: income_id,
            amount: income_request.amount,
            start_date: income_request.start_date,
            frequency: income_request.frequency
        }]);

        let budget_incomes = [];
        for (let date in budget_data) {
            for (let inc of budget_data[date]) {
                let budget_income = await db.pw_budget_table_income.create({
                    date: date,
                    user_id: user_id,
                    income_id: income_id,
                    source: inc.source,
                    amount: inc.amount,
                    start_date: inc.start_date,
                    frequency: inc.frequency
                });
                budget_incomes.push(budget_income);
            }
        }

        return { success: true, message: "Income and budget incomes created successfully", income: income };
    }

    async get_income(user_id, income_id) {
        const user = await db.pw_users.findOne({ where: { id: user_id } });
        if (!user) {
            return { success: false, message: "User not found" };
        }

        const income = await db.pw_income.findOne({ where: { id: income_id, user_id: user_id } });
        if (!income) {
            return { success: false, message: "Income not found or not owned by the user" };
        }

        return { success: true, message: "Income retrieved successfully", income: income };
    }

    async get_income_by_user(user_id) {
        const user = await db.pw_users.findOne({ where: { id: user_id } });
        if (!user) {
            return { success: false, message: "User not found" };
        }

        const incomes = await db.pw_income.findAll({ where: { user_id: user_id }, order: [['start_date', 'ASC']] });
        if (!incomes) {
            return { success: false, message: "Income not found or not owned by the user" };
        }

        return { success: true, message: "Income retrieved successfully", income: incomes };
    }

    async update_income(user_id, income_id, income_request) {
        const user = await db.pw_users.findOne({ where: { id: user_id } });
        if (!user) {
            return { success: false, message: "User not found" };
        }

        let income = await db.pw_income.findOne({ where: { id: income_id, user_id: user_id } });
        if (!income) {
            return { success: false, message: "Income not found or not owned by the user" };
        }

        income.source = income_request.source !== undefined ? income_request.source : income.source;
        income.amount = income_request.amount !== undefined ? income_request.amount : income.amount;
        income.start_date = income_request.start_date !== undefined ? income_request.start_date : income.start_date;
        income.frequency = income_request.frequency !== undefined ? income_request.frequency : income.frequency;

        await income.save();

        await db.pw_budget_table_income.destroy({ 
            where: { 
                date: { [Sequelize.Op.gt]: new Date() },
                income_id: income_id, 
                user_id: user_id 
            } 
        });

        let budget_data = group_by_date([{
            source: income_request.source,
            income_id: income_id,
            amount: income_request.amount,
            start_date: income_request.start_date,
            frequency: income_request.frequency
        }]);

        let budget_incomes = [];
        for (let date in budget_data) {
            if (new Date() > new Date(date)) {
                continue;
            }
            for (let inc of budget_data[date]) {
                let budget_income = await db.pw_budget_table_income.create({
                    date: date,
                    user_id: user_id,
                    income_id: income_id,
                    source: inc.source,
                    amount: inc.amount,
                    start_date: inc.start_date,
                    frequency: inc.frequency
                });
                budget_incomes.push(budget_income);
            }
        }

        return { success: true, message: "Income updated successfully", income: income };
    }

    async delete_income(user_id, income_id) {
        const user = await db.pw_users.findOne({ where: { id: user_id } });
        if (!user) {
            return { success: false, message: "User not found" };
        }

        const income = await db.pw_income.findOne({ where: { id: income_id, user_id: user_id } });
        if (!income) {
            return { success: false, message: "Income not found or not owned by the user" };
        }

        await db.pw_income.destroy({ where: { id: income_id } });

        await db.pw_budget_table_income.destroy({ 
            where: { 
                date: { [Sequelize.Op.gt]: new Date() },
                income_id: income_id, 
                user_id: user_id 
            } 
        });

        return { success: true, message: "Income deleted successfully" };
    }

}

module.exports = IncomeService;
